File: apps/core/app/src/test/java/com/beeboentertainment/movie/core/HomeTheaterTest.kt
package com.beeboentertainment.movie.core

import com.beeboentertainment.movie.data.NegotiateResponse
import com.beeboentertainment.movie.data.PlaybackInfo
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The declaration this TV / phone sends with POST /api/playback/negotiate and the rules for following the answer
 * (docs/HOME-THEATER.md). UNVERIFIED on a device: these prove the declaration and the decisions, not what a Shield or a
 * Fire TV really reports.
 */
class HomeTheaterTest {
    private val lenient = Json { ignoreUnknownKeys = true }

    private val uhdTv = DeviceCaps(
        client = "androidtv", name = "Living room Shield",
        video = mapOf(
            "h264" to VideoDecoderCaps(listOf("baseline", "main", "high"), listOf(8)),
            "hevc" to VideoDecoderCaps(listOf("main", "main10"), listOf(8, 10)),
            "vp9" to VideoDecoderCaps(listOf("profile0", "profile2"), listOf(8, 10)),
        ),
        maxHeight = 2160, hdr10 = true, hdr10Plus = false, hlg = true, dolbyVisionProfiles = listOf(5, 8), dvFallback = false,
        audio = mapOf(
            "aac" to AudioFormatCaps(decode = true, passthrough = false),
            "ac3" to AudioFormatCaps(decode = true, passthrough = true),
            "eac3" to AudioFormatCaps(decode = true, passthrough = true, atmos = true),
            "flac" to AudioFormatCaps(decode = true, passthrough = false),
        ),
        maxAudioChannels = 8
    )

    private fun JsonObject.audio(key: String): JsonObject? = this["audio"]?.jsonObject?.get(key)?.jsonObject

    @Test fun anUhdHdrShieldDeclaration() {
        val p = DeviceProfileBuilder.build(uhdTv)
        assertEquals(1, p["v"]!!.jsonPrimitive.int)
        assertEquals("androidtv", p["client"]!!.jsonPrimitive.content)
        assertEquals("Living room Shield", p["name"]!!.jsonPrimitive.content)
        assertEquals(setOf("h264", "hevc", "vp9"), p["video"]!!.jsonObject.keys)
        assertEquals(listOf(8, 10), p["video"]!!.jsonObject["hevc"]!!.jsonObject["bitDepths"]!!.jsonArray.map { it.jsonPrimitive.int })
        assertEquals(listOf("hdr10", "hlg", "dv:5,8"), p["hdr"]!!.jsonArray.map { it.jsonPrimitive.content })
        assertEquals(2160, p["maxHeight"]!!.jsonPrimitive.int)
        assertEquals(3840, p["maxWidth"]!!.jsonPrimitive.int)
        assertEquals(listOf("mp4", "mkv", "ts", "webm"), p["containers"]!!.jsonArray.map { it.jsonPrimitive.content })
        assertEquals(listOf("hls-ts", "hls-fmp4"), p["streaming"]!!.jsonArray.map { it.jsonPrimitive.content })
        assertEquals(listOf("vtt", "srt"), p["subtitles"]!!.jsonArray.map { it.jsonPrimitive.content })
        // it survives the trip over the wire
        assertEquals(p, lenient.parseToJsonElement(p.toString()))
    }

    @Test fun dolbyDigitalIsDecodedAndPassedThroughAndAtmosOnlyOnJoc() {
        val p = DeviceProfileBuilder.build(uhdTv)
        val eac3 = p.audio("eac3")!!
        assertTrue(eac3["passthrough"]!!.jsonPrimitive.boolean)
        assertTrue(eac3["atmos"]!!.jsonPrimitive.boolean)
        assertNull("decode is on by default: only a 'no' is written", eac3["decode"])
        assertNull(p.audio("ac3")!!["atmos"])
        assertEquals(6, p.audio("aac")!!["maxChannels"]!!.jsonPrimitive.int)
        assertEquals(8, p["maxAudioChannels"]!!.jsonPrimitive.int)
    }

    @Test fun trueHdAndDtsAreOffUnlessTheOutputReportedThem() {
        val p = DeviceProfileBuilder.build(uhdTv)
        for (key in listOf("truehd", "dts", "dtshd", "dtsx")) assertNull(key, p.audio(key))

        val withReceiver = uhdTv.copy(
            audio = uhdTv.audio + mapOf(
                "truehd" to AudioFormatCaps(decode = false, passthrough = true, atmos = true),
                "dts" to AudioFormatCaps(decode = false, passthrough = true),
                "dtshd" to AudioFormatCaps(decode = false, passthrough = true),
            )
        )
        val q = DeviceProfileBuilder.build(withReceiver)
        val truehd = q.audio("truehd")!!
        assertTrue(truehd["passthrough"]!!.jsonPrimitive.boolean)
        assertFalse(truehd["decode"]!!.jsonPrimitive.boolean)
        assertEquals(8, truehd["maxChannels"]!!.jsonPrimitive.int)
        assertNotNull(q.audio("dts"))
        assertNotNull(q.audio("dtshd"))
        assertNull("DTS:X is never claimed", q.audio("dtsx"))
    }

    @Test fun anSdrTvSaysSdrAndNoDolbyVision() {
        val sdr = uhdTv.copy(hdr10 = false, hlg = false, dolbyVisionProfiles = emptyList(), maxHeight = 1080)
        val p = DeviceProfileBuilder.build(sdr)
        assertEquals(JsonArray(emptyList()), p["hdr"])
        assertEquals(1080, p["maxHeight"]!!.jsonPrimitive.int)
        assertNull(p["maxWidth"])
        assertFalse(p.toString().contains("dv:"))
        assertFalse(p.toString().contains("hdr10plus"))
    }

    @Test fun hdr10PlusAndTheDolbyVisionFallbackAreOnlyWhatWasReported() {
        val p = DeviceProfileBuilder.build(uhdTv.copy(hdr10Plus = true, dolbyVisionProfiles = emptyList(), dvFallback = true))
        assertEquals(listOf("hdr10", "hdr10plus", "hlg", "dvfallback"), p["hdr"]!!.jsonArray.map { it.jsonPrimitive.content })
    }

    @Test fun aCodecWithNoDecoderIsAbsentAndNothingProbedIsStillValid() {
        val p = DeviceProfileBuilder.build(DeviceCaps(client = "android"))
        assertNull(p["video"])
        assertNull(p["audio"])
        assertEquals("android", p["client"]!!.jsonPrimitive.content)
        assertEquals(1080, p["maxHeight"]!!.jsonPrimitive.int)
        assertEquals(setOf("v", "client", "hdr", "maxHeight", "containers", "streaming", "subtitles"), p.keys)
    }

    @Test fun aTvWithTwoChannelSpeakersDoesNotClaimSurroundDecoding() {
        val speakers = uhdTv.copy(
            maxAudioChannels = 2,
            audio = mapOf(
                "aac" to AudioFormatCaps(decode = true, passthrough = false),
                "ac3" to AudioFormatCaps(decode = true, passthrough = false),
            )
        )
        val p = DeviceProfileBuilder.build(speakers)
        assertEquals(2, p.audio("aac")!!["maxChannels"]!!.jsonPrimitive.int)
        assertEquals(2, p["maxAudioChannels"]!!.jsonPrimitive.int)
    }

    @Test fun summaryIsAShortReadableLine() {
        assertEquals("h264 hevc vp9 · hdr10 hlg dv:5,8 · 2160p", DeviceProfileBuilder.summary(DeviceProfileBuilder.build(uhdTv)))
    }

    @Test fun onlyAServerWithTheRouteIsAskedAndOnlyForTheOriginal() {
        assertTrue(HomeTheaterRules.shouldNegotiate(serverHasRoute = true, wantsOriginal = true, casting = false, burningSubtitle = false))
        assertFalse("an older computer", HomeTheaterRules.shouldNegotiate(false, true, false, false))
        assertFalse("an explicit conversion is left alone", HomeTheaterRules.shouldNegotiate(true, false, false, false))
        assertFalse("the declaration describes this device, not the cast receiver", HomeTheaterRules.shouldNegotiate(true, true, true, false))
        assertFalse("a burnt-in subtitle needs the conversion the app already asks for", HomeTheaterRules.shouldNegotiate(true, true, false, true))
    }

    @Test fun onlyThePlansOnTheirOwnRoutesAreFollowed() {
        assertTrue(HomeTheaterRules.followable(PlayMethod.DIRECT_PLAY, "/file?id=a&mt=T"))
        assertTrue(HomeTheaterRules.followable(PlayMethod.DIRECT_PLAY, "/tvfile?id=a&mt=T"))
        assertTrue(HomeTheaterRules.followable(PlayMethod.DIRECT_STREAM, "/hls/T1/master.m3u8"))
        assertTrue(HomeTheaterRules.followable(PlayMethod.TRANSCODE, "/hls/T2/index.m3u8"))
        val bad = listOf(
            PlayMethod.DIRECT_PLAY to "http://evil.example/file?id=a",
            PlayMethod.DIRECT_PLAY to "//evil.example/file?id=a",
            PlayMethod.DIRECT_PLAY to "/api/admin/settings",
            PlayMethod.DIRECT_PLAY to "/hls/T/index.m3u8",
            PlayMethod.DIRECT_STREAM to "/file?id=a",
            PlayMethod.DIRECT_STREAM to "/hls/../x.m3u8",
            PlayMethod.DIRECT_STREAM to "/hls/T/evil.php",
            PlayMethod.TRANSCODE to "/x/index.m3u8",
            PlayMethod.DIRECT_PLAY to "/file?id=a\nX: y",
            PlayMethod.DIRECT_PLAY to "",
        )
        for ((m, u) in bad) assertFalse("$m $u", HomeTheaterRules.followable(m, u))
        assertFalse(HomeTheaterRules.followable(null, "/file?id=a"))
    }

    @Test fun thePreparingWaitIsBounded() {
        assertEquals(3_000L, HomeTheaterRules.prepareWaitMs(3.0))
        assertEquals(3_000L, HomeTheaterRules.prepareWaitMs(null))
        assertEquals(10_000L, HomeTheaterRules.prepareWaitMs(99.0))
        assertEquals(1_000L, HomeTheaterRules.prepareWaitMs(0.0))
    }

    @Test fun theAnswerIsParsedLeniently() {
        val plan = lenient.decodeFromString<NegotiateResponse>(
            """{"ok":true,"method":"DirectStream","url":"/hls/T9/master.m3u8","ticket":"T9","container":"hls-fmp4","durationSec":7020.5,"plan":{"reasonCodes":[]}}"""
        )
        assertEquals(PlayMethod.DIRECT_STREAM, plan.playMethod)
        assertTrue(plan.followable)
        assertEquals("T9", plan.ticket)
        val hostile = lenient.decodeFromString<NegotiateResponse>("""{"ok":true,"method":"DirectPlay","url":"http://evil.example/file?id=a"}""")
        assertFalse(hostile.followable)
        val notOk = lenient.decodeFromString<NegotiateResponse>("""{"ok":false,"method":"DirectPlay","url":"/file?id=a"}""")
        assertFalse(notOk.followable)
        val preparing = lenient.decodeFromString<NegotiateResponse>("""{"ok":false,"error":"preparing","retryAfterSec":3}""")
        assertEquals("preparing", preparing.error)
        assertNull(PlayMethod.fromWire("Teleport"))
    }

    @Test fun theInfoBlockIsTheFeatureTest() {
        val older = lenient.decodeFromString<PlaybackInfo>("""{"ok":true,"durationSec":10}""")
        assertNull(older.homeTheater)
        val newer = lenient.decodeFromString<PlaybackInfo>("""{"ok":true,"durationSec":10,"homeTheater":{"badges":["4K","Dolby Vision"],"plan":{}}}""")
        assertEquals(listOf("4K", "Dolby Vision"), newer.homeTheater!!.badges)
        val bare = lenient.decodeFromString<PlaybackInfo>("""{"ok":true,"homeTheater":{}}""")
        assertNotNull(bare.homeTheater)
    }
}

File: apps/smarttv/test/home-theater.test.mjs
import test from 'node:test'
import assert from 'node:assert/strict'
import fs from 'node:fs'
import path from 'node:path'
import { createRequire } from 'node:module'
import { fileURLToPath } from 'node:url'
import { buildDeviceProfile, clientNameFor, describeProfile, hdrList } from '../app/js/util/deviceProfile.js'
import { negotiateBody, normalizeNegotiate, prepareWaitSec, isPlayableRoute, describePlan, NEGOTIATE_PATH } from '../app/js/util/playback.js'
import { normalizePlaybackInfo } from '../app/js/util/models.js'
import { probeEnv, readDisplay, makeSupports } from '../app/js/platform/capabilities.js'
import { createClient } from '../app/js/api.js'
import { createStore } from '../app/js/store.js'

// Device profile (docs/HOME-THEATER.md): the honest declaration each web TV sends with POST /api/playback/negotiate.

// A fake engine: it says yes to a plain mime type that equals one of the given words, and to a codec-bearing type that
// contains one of them ("video/mp4" alone does not make every mp4 codec playable).
const engine = (...yes) => (mime) => yes.some((f) => (f.includes('/') ? mime === f : mime.includes('codecs=') && mime.includes(f)))
const UHD_TV_ENGINE = engine(
  'avc1.640028', 'avc1.640029', 'avc1.640033', 'avc1.640034', 'avc1.4d401f', 'avc1.42e01e',
  'hvc1.1.6.L120.90', 'hvc1.1.6.L153.90', 'hvc1.2.4.L153.B0', 'vp09.00.10.08', 'vp09.02.10.10', 'av01.0.05M.08', 'av01.0.05M.10',
  'mp4a.40.2', 'ac-3', 'ec-3', 'audio/flac', 'opus', 'audio/mpeg',
  'video/mp4', 'video/x-matroska', 'video/webm', 'video/mp2t', 'application/vnd.apple.mpegurl'
)
const HD_TV_ENGINE = engine('avc1.640028', 'avc1.640029', 'avc1.4d401f', 'avc1.42e01e', 'mp4a.40.2', 'ac-3', 'video/mp4', 'application/x-mpegURL')

test('client names are the ones the server knows', () => {
  assert.equal(clientNameFor('tizen'), 'samsung')
  assert.equal(clientNameFor('webos'), 'lg')
  assert.equal(clientNameFor('xbox'), 'xbox')
  assert.equal(clientNameFor('browser'), 'chrome')
  assert.equal(clientNameFor('anything'), 'chrome')
})

test('nothing to probe -> a bare { v, client }: the server keeps its own default for the platform', () => {
  assert.deepEqual(buildDeviceProfile({ platform: 'tizen' }), { v: 1, client: 'samsung' })
  assert.deepEqual(buildDeviceProfile(null), { v: 1, client: 'chrome' })
  assert.equal(buildDeviceProfile({ platform: 'webos', name: 'Bad <name>' }).name, 'Bad name')
})

test('Samsung UHD HDR TV: only what the engine and the display reported', () => {
  const p = buildDeviceProfile({
    platform: 'tizen', name: 'Beebo TV app', supports: UHD_TV_ENGINE, hlsNative: true, hlsJs: false,
    display: { uhd: true, hdr10: true, hdr10plus: null, hlg: null, dvProfiles: null, atmos: null }
  })
  assert.equal(p.client, 'samsung')
  assert.deepEqual(Object.keys(p.video).sort(), ['av1', 'h264', 'hevc', 'vp9'])
  assert.deepEqual(p.video.hevc.bitDepths, [8, 10])
  assert.equal(p.video.h264.maxLevel, 52)
  assert.deepEqual(p.hdr, ['hdr10'])
  assert.equal(p.maxHeight, 2160)
  assert.equal(p.maxWidth, 3840)
  assert.ok(p.containers.includes('mkv') && p.containers.includes('mp4'))
  // Fragmented-MP4 HLS cannot be asked of a native player: TS pieces only.
  assert.deepEqual(p.streaming, ['hls-ts'])
  assert.deepEqual(p.subtitles, ['vtt'])
  assert.equal(p.audio.eac3.atmos, undefined)
  assert.equal(p.maxAudioChannels, 6)
})

test('a TV never claims what the platform cannot tell: no HDR10+, no Dolby Vision, no DTS / TrueHD, no passthrough', () => {
  // an engine that says YES to everything, and a display that reports nothing
  const p = buildDeviceProfile({ platform: 'tizen', supports: () => true, hlsNative: true, display: { uhd: null, hdr10: null, hdr10plus: null, hlg: null, dvProfiles: null, atmos: null } })
  assert.deepEqual(p.hdr, [])
  assert.equal(p.maxHeight, 1080, 'unknown panel size = 1080')
  for (const k of ['truehd', 'dts', 'dtshd', 'dtsx']) assert.equal(p.audio[k], undefined, k)
  for (const a of Object.values(p.audio)) { assert.equal(a.passthrough, undefined); assert.equal(a.atmos, undefined) }
  assert.ok(!JSON.stringify(p).includes('dv:'))
  assert.ok(!JSON.stringify(p).includes('hdr10plus'))
})

test('a plain HD TV: no HEVC, no HDR, 1080p', () => {
  const p = buildDeviceProfile({ platform: 'webos', supports: HD_TV_ENGINE, hlsNative: true, display: { uhd: false, hdr10: false, dvProfiles: [] } })
  assert.deepEqual(Object.keys(p.video), ['h264'])
  assert.deepEqual(p.hdr, [])
  assert.equal(p.maxHeight, 1080)
  assert.equal(p.maxWidth, undefined)
  assert.deepEqual(p.audio.ac3, { maxChannels: 6 })
  assert.equal(p.audio.eac3, undefined)
  assert.deepEqual(p.containers, ['mp4'])
})

test('LG webOS reporting HDR10, Dolby Vision and Atmos', () => {
  const p = buildDeviceProfile({
    platform: 'webos', supports: UHD_TV_ENGINE, hlsNative: true,
    display: { uhd: true, hdr10: true, hlg: true, dvProfiles: [5, 8], atmos: true }
  })
  assert.deepEqual(p.hdr, ['hdr10', 'hlg', 'dv:5,8'])
  assert.deepEqual(p.audio.eac3, { maxChannels: 8, atmos: true })
  assert.equal(p.maxAudioChannels, 8)
})

test('Xbox playing HLS through hls.js may say fragmented-MP4 HLS; native HLS only claims TS', () => {
  const viaJs = buildDeviceProfile({ platform: 'xbox', supports: UHD_TV_ENGINE, hlsJs: true, hlsNative: false, display: { uhd: true, hdr10: true } })
  assert.deepEqual(viaJs.streaming, ['hls-ts', 'hls-fmp4'])
  assert.equal(viaJs.client, 'xbox')
  const native = buildDeviceProfile({ platform: 'xbox', supports: engine('avc1.640028', 'video/mp4'), hlsNative: true, hlsJs: false })
  assert.deepEqual(native.streaming, ['hls-ts'])
  const none = buildDeviceProfile({ platform: 'xbox', supports: engine('avc1.640028', 'video/mp4'), hlsNative: false, hlsJs: false })
  assert.equal(none.streaming, undefined)
})

test('an engine that throws is treated as "no"', () => {
  const p = buildDeviceProfile({ platform: 'tizen', supports: () => { throw new Error('boom') } })
  assert.equal(p.video, undefined)
  assert.equal(p.audio, undefined)
  assert.deepEqual(p.hdr, [])
})

test('hdrList and describeProfile', () => {
  assert.deepEqual(hdrList({ hdr10: true, hlg: true, dvProfiles: [8] }), ['hdr10', 'hlg', 'dv:8'])
  assert.deepEqual(hdrList(null), [])
  assert.equal(describeProfile(null), '')
  assert.match(describeProfile({ video: { hevc: {}, h264: {} }, hdr: ['hdr10'], audio: { eac3: {} }, maxHeight: 2160 }), /hevc h264 · hdr10 · eac3 · 2160p/)
})

// ---- the server really reads it the way we mean (skipped when the desktop tree is not next to this app) --------------

const here = path.dirname(fileURLToPath(import.meta.url))
const serverModule = path.resolve(here, '../../../desktop/apps/desktop/electron/deviceProfile.js')
const haveServer = fs.existsSync(serverModule)
const serverProfile = haveServer ? createRequire(import.meta.url)(serverModule) : null

test('the desktop server parses our declaration and overrides only what we stated', { skip: !haveServer }, () => {
  const p = buildDeviceProfile({
    platform: 'tizen', supports: UHD_TV_ENGINE, hlsNative: true,
    display: { uhd: true, hdr10: true }
  })
  const r = serverProfile.resolveProfile({ declared: p })
  assert.equal(r.client, 'samsung')
  assert.equal(r.hdr.hdr10, true)
  assert.equal(r.hdr.hdr10plus, false, 'Samsung default assumes HDR10+; we did not claim it, so it is off')
  assert.deepEqual(r.hdr.dv, [])
  assert.deepEqual(r.streaming, ['hls-ts'])
  assert.ok(r.containers.includes('mkv'))
  assert.equal(r.audio.truehd, undefined)
  assert.equal(r.audio.dtshd, undefined)
  assert.equal(r.maxHeight, 2160)
  assert.equal(r.source, 'declared')
  // a bare declaration is just the platform default
  const bare = serverProfile.resolveProfile({ declared: buildDeviceProfile({ platform: 'webos' }) })
  assert.equal(bare.client, 'lg')
  assert.match(bare.source, /default:lg|declared\+default:lg/)
})

test('an SDR-only declaration (unknown display) turns the platform default HDR off on the server', { skip: !haveServer }, () => {
  const p = buildDeviceProfile({ platform: 'webos', supports: UHD_TV_ENGINE, hlsNative: true, display: {} })
  const r = serverProfile.resolveProfile({ declared: p })
  assert.equal(r.hdr.hdr10, false)
  assert.deepEqual(r.hdr.dv, [])
  assert.equal(r.maxHeight, 1080)
})

// ---- the probe (fake windows) ------------------------------------------------------------------------------------------

function fakeWindow(extra) {
  const video = { canPlayType: (m) => (UHD_TV_ENGINE(m) ? 'maybe' : '') }
  return { document: { createElement: (t) => (t === 'video' ? video : {}) }, ...extra }
}

test('probeEnv reads Tizen product info and HDR support', async () => {
  const win = fakeWindow({ webapis: { productinfo: { isUdPanelSupported: () => true }, avinfo: { isHdrTvSupport: () => true } } })
  const env = await probeEnv(win, 'tizen', { name: 'TV' })
  assert.equal(env.display.uhd, true)
  assert.equal(env.display.hdr10, true)
  assert.equal(env.hlsJs, false)
  assert.equal(env.supports('audio/mp4; codecs="ec-3"'), true)
  assert.equal(env.supports('audio/mp4; codecs="dtsc"'), false)
})

test('probeEnv reads webOS.deviceInfo, and gives up (unknown) if it never answers', async () => {
  const win = fakeWindow({ webOS: { deviceInfo: (cb) => cb({ uhd: 'true', hdr10: true, dolbyVision: true, dolbyAtmos: false }) } })
  const d = await readDisplay(win, 'webos')
  assert.deepEqual(d, { uhd: true, hdr10: true, hdr10plus: null, hlg: true, dvProfiles: [5, 8], atmos: false })
  const never = await readDisplay({ webOS: { deviceInfo: () => {} } }, 'webos')
  assert.equal(never.hdr10, null)
  const missing = await readDisplay({}, 'webos')
  assert.equal(missing.uhd, null)
})

test('probeEnv on a Chromium web view: HDR from the media query, UHD from the physical screen, hls.js from the engine hook', async () => {
  const win = fakeWindow({ matchMedia: (q) => ({ matches: q === '(dynamic-range: high)' }), screen: { width: 1920, height: 1080 }, devicePixelRatio: 2 })
  const env = await probeEnv(win, 'xbox', { engine: () => 'hlsjs' })
  assert.equal(env.display.hdr10, true)
  assert.equal(env.display.uhd, true)
  assert.equal(env.hlsJs, true)
  assert.equal(env.hlsNative, false)
  const sdr = await readDisplay({ matchMedia: () => ({ matches: false }), screen: { width: 1920, height: 1080 }, devicePixelRatio: 1 }, 'xbox')
  assert.equal(sdr.hdr10, false)
  assert.equal(sdr.uhd, false)
})

test('makeSupports uses canPlayType, then MediaSource for codec types only', () => {
  const video = { canPlayType: (m) => (m === 'video/mp4' ? 'probably' : '') }
  const win = { MediaSource: { isTypeSupported: (m) => m.includes('hvc1') } }
  const s = makeSupports(win, video)
  assert.equal(s('video/mp4'), true)
  assert.equal(s('video/mp4; codecs="hvc1.2.4.L153.B0"'), true)
  assert.equal(s('video/x-matroska'), false)
  assert.equal(makeSupports({}, null)('video/mp4'), false)
})

// ---- negotiate: body, answer, client -------------------------------------------------------------------------------------

test('negotiateBody carries the profile in the body (a TV app cannot send extra headers cross-origin)', () => {
  const profile = { v: 1, client: 'samsung' }
  const b = negotiateBody({ kind: 'tv', id: 'ep1', client: 'samsung', profile, audio: 2 })
  assert.deepEqual(b, { kind: 'tv', id: 'ep1', client: 'samsung', deviceProfile: profile, quality: 'original', audio: 2 })
  assert.equal(negotiateBody({ kind: 'weird', id: 'm', quality: '720p' }).kind, 'movie')
  assert.equal(negotiateBody({ kind: 'movie', id: 'm', quality: '720p' }).quality, '720p')
  assert.equal(negotiateBody({ kind: 'movie', id: 'm' }).deviceProfile, undefined)
})

const PLAN = { reasonCodes: ['CONTAINER_NOT_SUPPORTED'], playsAs: '4K HDR10 (copied)' }

test('normalizeNegotiate accepts exactly the three plans and their own routes', () => {
  const dp = normalizeNegotiate({ ok: true, method: 'DirectPlay', url: '/file?id=abc&mt=TOK', mimeType: 'video/x-matroska', durationSec: 6300, plan: PLAN })
  assert.equal(dp.method, 'DirectPlay')
  assert.equal(dp.url, '/file?id=abc&mt=TOK')
  assert.equal(dp.durationSec, 6300)
  assert.deepEqual(dp.reasons, ['CONTAINER_NOT_SUPPORTED'])
  assert.equal(normalizeNegotiate({ ok: true, method: 'DirectPlay', url: '/tvfile?id=e&mt=T' }).method, 'DirectPlay')
  const ds = normalizeNegotiate({ ok: true, method: 'DirectStream', url: '/hls/T1.abc-9/master.m3u8', ticket: 'T1.abc-9', container: 'hls-fmp4' })
  assert.equal(ds.method, 'DirectStream')
  assert.equal(ds.ticket, 'T1.abc-9')
  const tc = normalizeNegotiate({ ok: true, method: 'Transcode', url: '/hls/T2/index.m3u8', ticket: 'T2' })
  assert.equal(tc.method, 'Transcode')
  assert.equal(describePlan(dp), 'Direct play')
  assert.equal(describePlan(ds), 'Direct stream')
  assert.equal(describePlan(tc), 'Converted')
  assert.equal(describePlan(null), '')
})

test('normalizeNegotiate refuses a hostile or unknown answer (the caller falls back to /playback/start)', () => {
  const bad = [
    null, 5, 'x', {}, { ok: false, method: 'DirectPlay', url: '/file?id=a' },
    { ok: true, method: 'Teleport', url: '/file?id=a' },
    { ok: true, method: 'DirectPlay', url: 'http://evil.example/file?id=a' },
    { ok: true, method: 'DirectPlay', url: '//evil.example/file?id=a' },
    { ok: true, method: 'DirectPlay', url: '/api/admin/settings' },
    { ok: true, method: 'DirectPlay', url: '/hls/T/index.m3u8' },
    { ok: true, method: 'DirectStream', url: '/file?id=a&mt=T' },
    { ok: true, method: 'DirectStream', url: '/hls/T/../../login' },
    { ok: true, method: 'DirectStream', url: '/hls/T/evil.php' },
    { ok: true, method: 'Transcode', url: '/x.m3u8' },
    { ok: true, method: 'Transcode' }
  ]
  for (const b of bad) assert.equal(normalizeNegotiate(b), null, JSON.stringify(b))
  assert.equal(isPlayableRoute('DirectPlay', '/file?id=1'), true)
  assert.equal(isPlayableRoute('DirectStream', '/hls/a/master.m3u8'), true)
  assert.equal(isPlayableRoute('DirectStream', '/hls/a/master.m3u8\n'), false)
})

test('prepareWaitSec: only a 503 "preparing" is worth asking again', () => {
  assert.equal(prepareWaitSec({ status: 503, body: { error: 'preparing', retryAfterSec: 3 } }), 3)
  assert.equal(prepareWaitSec({ status: 503, body: { error: 'preparing', retryAfterSec: 99 } }), 10)
  assert.equal(prepareWaitSec({ status: 503, body: { error: 'preparing' } }), 3)
  assert.equal(prepareWaitSec({ status: 503, body: { error: 'other' } }), 0)
  assert.equal(prepareWaitSec({ status: 404, body: { error: 'preparing' } }), 0)
  assert.equal(prepareWaitSec(null), 0)
})

test('normalizePlaybackInfo: the homeTheater block is the feature test for the negotiate route', () => {
  assert.equal(normalizePlaybackInfo({ durationSec: 10 }).homeTheater, false)
  const i = normalizePlaybackInfo({ durationSec: 10, homeTheater: { badges: ['4K', 'Dolby Vision', '<b>x</b>', '', 'Atmos'], plan: {} } })
  assert.equal(i.homeTheater, true)
  assert.deepEqual(i.badges, ['4K', 'Dolby Vision', '<b>x</b>', 'Atmos']) // inert text: views use textContent
  assert.deepEqual(normalizePlaybackInfo({ homeTheater: 'yes' }).badges, [])
})

function fakeXHR(handler) {
  const seen = []
  class X {
    constructor() { this.headers = {}; this.status = 0; this.responseText = '' }
    open(method, url) { this.method = method; this.url = url }
    setRequestHeader(k, v) { this.headers[k] = v }
    send(body) {
      seen.push(this)
      this.body = body
      const r = handler(this)
      setTimeout(() => { this.status = r.status; this.responseText = JSON.stringify(r.body); this.onload() }, 0)
    }
  }
  X.seen = seen
  return X
}

test('client.playbackNegotiate posts the declared profile with the bearer token, and parses the plan', async () => {
  const profile = { v: 1, client: 'samsung', hdr: [] }
  const XHR = fakeXHR(() => ({ status: 200, body: { ok: true, method: 'DirectPlay', url: '/file?id=m1&mt=T', plan: PLAN } }))
  const client = createClient({ XHR, getOrigin: () => 'http://h:47811', getToken: () => 'TOK', getProfile: () => profile, clientName: 'samsung' })
  const plan = await client.playbackNegotiate('movie', 'm1', { quality: 'original', audio: 3 })
  assert.equal(plan.method, 'DirectPlay')
  const x = XHR.seen[0]
  assert.equal(x.method, 'POST')
  assert.equal(x.url, 'http://h:47811' + NEGOTIATE_PATH)
  assert.equal(x.headers.Authorization, 'Bearer TOK')
  assert.equal(x.headers['X-Beebo-Device-Profile'], undefined, 'the profile is in the body, not in a header')
  assert.equal(x.headers['X-Beebo-Client'], undefined)
  assert.deepEqual(JSON.parse(x.body), { kind: 'movie', id: 'm1', client: 'samsung', deviceProfile: profile, quality: 'original', audio: 3 })
  assert.ok(!x.url.includes('TOK'))
})

test('client.playbackNegotiate: an unusable answer is bad_response, "preparing" keeps its body', async () => {
  const bad = createClient({ XHR: fakeXHR(() => ({ status: 200, body: { ok: true, method: 'DirectPlay', url: 'http://evil/x' } })), getOrigin: () => 'http://h', getToken: () => 't' })
  await assert.rejects(bad.playbackNegotiate('movie', 'a'), (e) => e.kind === 'bad_response')
  const prep = createClient({ XHR: fakeXHR(() => ({ status: 503, body: { ok: false, error: 'preparing', retryAfterSec: 3 } })), getOrigin: () => 'http://h', getToken: () => 't' })
  await assert.rejects(prep.playbackNegotiate('movie', 'a'), (e) => prepareWaitSec(e) === 3)
})

test('store: "play original" defaults on and can be switched off', () => {
  const mem = new Map()
  const storage = { getItem: (k) => (mem.has(k) ? mem.get(k) : null), setItem: (k, v) => mem.set(k, v), removeItem: (k) => mem.delete(k) }
  const s = createStore(storage)
  assert.equal(s.getPlayOriginal(), true)
  s.setPlayOriginal(false)
  assert.equal(s.getPlayOriginal(), false)
  s.setPlayOriginal(true)
  assert.equal(s.getPlayOriginal(), true)
})
